ToricVarieties: Additions for divisors, divisor classes and tensor products for line bundles

HereAround opened the pull request:
As requested by @simontelen, this PR aims towards closing https://github.com/oscar-system/Oscar.jl/issues/1036:

- Change names of maps between characters, (Cartier) toric divisors, class group and Picard group.
- Support addition, subtraction and scalar multiples of toric divisors.
- Overload equality for toric divisors.
- Introduce divisor classes and support their addition, subtraction and scalar multiples.
- Overload equality for toric divisor classes.
- Support tensor products, tensor powers and inverses of toric line bundles.
- Overload equality for toric line bundles.
- Add new constructor for toric line bundle from toric divisor.
- Change `variety(ToricLineBundle)` to `toric_variety(ToricLineBundle)` for consistency.

Since this PR changes the source files for toric line bundles, there is mild tension (as in rebase necessary) with https://github.com/oscar-system/Oscar.jl/pull/1054.

cc @lkastner

HereAround committed:
ToricVarieties: map_from_character_to_principal_divisors -> map_from_char_to_divT

HereAround committed:
ToricVarieties: map_from_weil_divisors_to_class_group -> map_from_divT_to_cl

HereAround committed:
ToricVarieties: map_from_cartier_divisor_group_to_torus_invariant_divisor_group -> map_from_torus_invariant_cartier_divisor_group_to_torus_invariant_weil_divisor_group

HereAround committed:
ToricVarieties: map_from_cartier_divisor_group_to_picard_group -> map_from_cdivT_to_pic

HereAround committed:
ToricVarieties: Add addition of toric divisors

HereAround committed:
ToricVarieties: Add scalar multiplication of toric divisors

HereAround committed:
ToricVarieties: Add difference of two toric divisors

HereAround committed:
ToricVarieties: Implement toric divisor classes

HereAround committed:
ToricVarieties: Add equality of toric divisors

HereAround committed:
ToricVarieties: Support tensor products and equality of line bundles

HereAround committed:
ToricVarieties: Change variety(line_bundle) -> toric_variety(line_bundle)

HereAround committed:
ToricVarieties: Add constructor for toric line bundle associated to toric divisor

fingolfin commented on docs/src/ToricVarieties/NormalToricVarieties.md:
```diff
@@ -86,10 +86,10 @@ affine_open_covering( v::AbstractNormalToricVariety )
+map_from_cdivT_to_divT(v::AbstractNormalToricVariety)
+map_from_cdivT_to_pic(v::AbstractNormalToricVariety)
+map_from_char_to_divT(v::AbstractNormalToricVariety)
+map_from_divT_to_cl(v::AbstractNormalToricVariety)
```
All these renamings are breaking changes, so if we merge them we have to increase the version to 0.9.0. alternatively use `@deprecate`?


Also, these names are kinda against our conventions. But if they are internal functions only, it should be fine

HereAround replied:
Just so I understand: When you say against the conventions, then you complain about the capital "T"?

I do not understand the point about `breaking changes`. Could you please elaborate?

lkastner replied:
I hope I am explaining the right part, but @fingolfin just means that someone could already be using Oscar.jl in the current version and then they might also use these functions with their current names. So we have to make sure their code still works with this current version of Oscar.jl, even after we change the names. We can do that by bumping the version number, then their now old version is still ok. Or we introduce the old functions as shortcuts marked by the `@deprecate` macro(?). Maybe the latter even makes them aware of the change.

If I understand it correctly, then the syntax is
```
@deprecate old_function(args::ArgsType) new_function(args)
```
Maybe we can try it out here?
```
@deprecate map_from_cartier_divisor_group_to_torus_invariant_divisor_group(v::AbstractNormalToricVariety) map_from_cdivT_to_divT(v)
@deprecate map_from_cartier_divisor_group_to_picard_group(v::AbstractNormalToricVariety) map_from_cdivT_to_pic(v)
[...]
```
I would place these just at the end of the central `JToric.jl`. Did I understand correctly @fingolfin?

thofma replied:
Why do we want to remove these perfectly fine function names with things that are hard to read/parse? I think we are moving away from `idk_wt_divT_or_cdivT_is` :)

HereAround replied:
Thank you for elaborating @lkastner.

@thofma The rational for this renaming is to make closer contact with the literature/book. In this particular case, the toric geometry book by Cox-Little-Schenk. In there, the group of torus invariant Weil divisors is denoted as Div_T and that of Cartier torus invariant Weil divisors by CDiv_T. (And similarly Cl for class group and Pic for Picard group.) The hope would be that the new names are easier/more convenient to use.

Indeed, I observed myself that the current function names are somewhat cumbersome. This in turn fits with the observation of @simontelen who brought this issue up in a discussion Monday last week.

lkastner replied:
> Why do we want to remove these perfectly fine function names with things that are hard to read/parse? I think we are moving away from `idk_wt_divT_or_cdivT_is` :)

Oh, the reason in this case was just that these were getting really long and we thought about using the notation from the book. If this is discouraged, we will refrain from doing so.

fingolfin replied:
@lkastner you explained it very well, thank you.

But my primary concern is that what @thofma said: in general we are moving towards more self-explanatory names for functions: It is clear to me what `map_from_cartier_divisor_group_to_picard_group` does without looking at the code. Can't say the same for `map_from_cdivT_to_pic`...

You don't really explain the motivation for this; I've just also asked a related question on issue #1036.

The main issue I see with `map_from_cartier_divisor_group_to_picard_group` is that it is really, really long.

thofma replied:
If I understand it correctly, we try to use the mathematical name (which is hopefully standard in the community) and not the mathematical notation used in a particular reference. As an example, we don't have `cl` instead of `class_group`. I know there are some exceptions in Oscar land, but I wish we would not add more.

Maybe this is another case where we would need aliases? Not sure though.

wdecker replied:
I support in general longer names, for example primary_decomposition over primdec. However, if there are well-established shortcuts, known even to beginners in the area, why not use them? For example, everybody is happy with using ext or would someone prefer to replace this by extension_group or some such. The same should apply to pic etc.

lkastner commented:
So as I said I have code for the addition of divisors etc. I just throw this away now, right? 😜

HereAround commented:
@lkastner: I failed to notice that you were already working on this. Sorry. (Otherwise, I would not have worked on this of course...)

What exactly does your code cover? Additions/scalar multiples of toric Weil divisors? Do you also have the following?
- Divisor classes (as in elements of cl(X)) including some simple constructors, properties, additions and scalar multiples.
- Tensor products, powers, inverses of line bundles (i.e. covering Pic(x)).
- Equality of toric Weil divisors, divisor classes and line bundles.

I have also made two minor changes to the line bundles:
- Rename the attribute `variety` of a line bundle into `toric_variety` (for consistency with terminology with toric (Weil) divisors).
- Added a new constructor for line bundles from toric Cartier divisor.

Could you maybe open a draft so I can take a look at your approach (or provide a link if I failed to notice it)? Maybe an efficient way to go about this is then that I remove the parts that your code covers and focus on the rest?

(Turned into draft for obvious reasons. And, for all it matters, errors fixed.)

lkastner commented on src/ToricVarieties/ToricDivisorClasses/attributes.jl:
```diff
@@ -0,0 +1,73 @@
+"""
+function class(tdc::ToricDivisorClass)
+    return tdc.class
+end
```
Maybe we want to make this such a one-liner?
```suggestion
class(tdc::ToricDivisorClass) = tdc.class
```

HereAround replied:
Thank you.

lkastner commented on src/ToricVarieties/ToricDivisorClasses/attributes.jl:
```diff
@@ -0,0 +1,73 @@
+"""
+function toric_variety(tdc::ToricDivisorClass)
+    return tdc.toric_variety
+end
```
Same here?
```suggestion
toric_variety(tdc::ToricDivisorClass) = tdc.toric_variety
```

lkastner reviewed (approved):
There are a bunch of functions that can be shortened from 3 to 1 line, all of the pattern
```
function foo(bar)
    return something
end
```
could become
```
foo(bar) = seomething
```
For example many of the `Base.{+,-,*}` are of this type. This is not super urgent as I think both variants are readable. And I also don't know whether there is any advantage to one over the other.

HereAround commented:
> There are a bunch of functions that can be shortened from 3 to 1 line, all of the pattern
> ```
> function foo(bar)
>     return something
> end
> ```
> could become
> ```
> foo(bar) = seomething
> ```
> For example many of the `Base.{+,-,*}` are of this type. This is not super urgent as I think both variants are readable. And I also don't know whether there is any advantage to one over the other.

I will look into this.

HereAround committed:
ToricVarieties: Simple functions to one-liners

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

HereAround commented:
I have just rebased this and PR made a few more changes:
- Turn simple functions into one-liners.
- Used `iszero()` to tell if an element of the Picard group or the class group is trivial. Thereby, we can decide equality of divisor classes and line bundles.
- I cache `istrivial` for divisor classes. It seems to me that this property is not cached for elements of f.p. Abelian groups.
- Added @deprecate for the function whose name was changed.

lkastner commented:
@fingolfin @HereAround On Friday @fieker asked why `class_group` in our case would not return "the map" as well. Here is the help string:
```
help?> class_group
search: class_group ray_class_group locally_free_class_group

  class_group(K::AnticNumberField) -> GrpAbFinGen, Map

  Shortcut for class_group(maximal_order(K)): returns the class group as an abelian group
  and a map from this group to the set of ideals of the maximal order.

  ──────────────────────────────────────────────────────────────────────────────────────────

  class_group(O::NfOrd; bound = -1, method = 3, redo = false, large = 1000) -> GrpAbFinGen, Map

  Returns a group A and a map f from A to the set of ideals of O. The inverse of the map is
  the projection onto the group of ideals modulo the group of principal ideals. redo allows
  to trigger a re-computation, thus avoiding the cache. bound, when given, is the bound for
  the factor base.
```
I am still thinking about this, but maybe we can go in the same direction?

HereAround commented:
I find this counter-intuitive: When I ask for `class_group`, then I would expect a group in return.

Besides, is there really a canonical map to the `class_group` in the toric setting? At least the following three come to my mind:
1. Div_T -> Cl
2. CDiv_T -> Cl
3. Pic -> Cl

Which of those is "the" map and why?

HereAround committed:
ToricVarieties: cartier_divisor_group -> torus_invariant_cartier_divisor_group

HereAround committed:
ToricVarieties: torusinvariant_divisor_group -> torus_invariant_weil_divisor_group

HereAround committed:
ToricVarieties: torus_invariant -> torusinvariant

HereAround commented:
I have just rebased this PR. Crucially, I have adjusted the new names of the relevant functions/groups. I hope to have taken all the input from the discussion above and in https://github.com/oscar-system/Oscar.jl/issues/1036 into account.

My proposal are the following names:

    torus_invariant_weil_divisor_group,
    torus_invariant_cartier_divisor_group,
    class_group,
    picard_group,
    map_from_character_lattice_to_torus_invariant_weil_divisor_group,
    map_from_torus_invariant_weil_divisor_group_to_class_group,
    map_from_torus_invariant_cartier_divisor_group_to_torus_invariant_weil_divisor_group,
    map_from_torus_invariant_cartier_divisor_group_to_picard_group.

The old names are marked with @ deprecated.

I hope that people are happy with this change. If yes, please consider merging.

HereAround commented:
@lkastner @fieker @fingolfin 

I believe it would be good to discuss the return value of `class_group` in a separate issue (and its possible resolution in a separate PR). I have opened a corresponding issue here: https://github.com/oscar-system/Oscar.jl/issues/1095

fingolfin commented on src/ToricVarieties/ToricDivisorClasses/constructors.jl:
```diff
@@ -0,0 +1,208 @@
+"""
+function Base.:+(tdc1::ToricDivisorClass, tdc2::ToricDivisorClass)
+    # check input
+    if !(toric_variety(tdc1) === toric_variety(tdc2))
```
Just as a general remark (no need to change this PR), you can also write this:
```suggestion
    if toric_variety(tdc1) !== toric_variety(tdc2)
```

fingolfin commented on src/ToricVarieties/ToricDivisorClasses/properties.jl:
````diff
@@ -0,0 +1,23 @@
+false
+```
+"""
+function istrivial(tdc::ToricDivisorClass)
+    return get_attribute!(tdc, :istrivial) do
+        return iszero(divisor_class(tdc))::Bool
+    end
+end
````
```suggestion
@attr Bool istrivial(tdc::ToricDivisorClass) = iszero(divisor_class(tdc))
```

fingolfin commented on src/ToricVarieties/ToricDivisorClasses/constructors.jl:
```diff
@@ -0,0 +1,208 @@
+#########################
+
+@attributes mutable struct ToricDivisorClass
+            toric_variety::AbstractNormalToricVariety
```
That's a weird indentation; we usually do 2/3/4 spaces, but this looks like it tries to align things with the start of the word "mutable" which is unusual

fingolfin commented on src/ToricVarieties/ToricDivisorClasses/attributes.jl:
````diff
@@ -0,0 +1,69 @@
+```
+"""
+function toric_divisor(tdc::ToricDivisorClass)
+    return get_attribute!(tdc, :toric_divisor) do
````
Could use `@attr` here, too

fingolfin commented on src/ToricVarieties/ToricLineBundles/constructors.jl:
````diff
@@ -48,7 +48,112 @@ end
+```
+"""
+function ToricLineBundle(v::AbstractNormalToricVariety, d::ToricDivisor)
+    if !(iscartier(d))
````
```suggestion
    if !iscartier(d)
```

fingolfin reviewed (approved):
Fine by me. I've left some suggestions, which you may or may not want to handle in a follow-up PR :-)
